Sources/AsyncAlgorithms/AsyncShareSequence.swift

```diff
@@ -0,0 +1,455 @@
+    
+    func emit(_ result: Result<Element?, Failure>) {
+      let (resumptions, demandContinuation) = state.withLock { state -> ([Resumption], CheckedContinuation<Void, Never>?) in
+        var resumptions = [Resumption]()
+        switch result {
+        case .success(let element):
+          if let element {
+            state.enqueue(element)
+          } else {
+            state.finished = true
+          }
+        case .failure(let failure):
+          state.finished = true
+          state.failure = failure
+        }
+        for (id, side) in state.sides {
+          if let continuation = side.continuaton {
+            if side.position < state.buffer.count {
+              resumptions.append(Resumption(continuation: continuation, result: .success(state.buffer[side.position])))
+              state.sides[id]?.position += 1
+              state.sides[id]?.continuaton = nil
+            } else if state.finished {
+              state.sides[id]?.continuaton = nil
+              if let failure = state.failure {
+                resumptions.append(Resumption(continuation: continuation, result: .failure(failure)))
+              } else {
+                resumptions.append(Resumption(continuation: continuation, result: .success(nil)))
+              }
+            }
+          }
+        }
+        state.trimBuffer()
+        if let limit, state.buffer.count < limit {
+          defer {
+            state.demand = nil
+          }
+          return (resumptions, state.demand)
+        } else {
+          return (resumptions, nil)
+        }
+      }
+      if let demandContinuation {
+        demandContinuation.resume()
+      }
+      for resumption in resumptions {
+        resumption.resume()
+      }
+    }
+    
+    func next(isolation actor: isolated (any Actor)?, id: Int) async throws(Failure) -> Element? {
+      let (base, cancelled) = state.withLock { state -> (Base?, Bool) in
+        switch state.iteratingTask {
+        case .pending(let base):
+          state.iteratingTask = .starting
+          return (base, false)
+        case .cancelled:
+          return (nil, true)
+        default:
+          return (nil, false)
+        }
+      }
+      if cancelled { return nil }
+      if let base {
+        nonisolated(unsafe) let transfer = base.makeAsyncIterator()
+        let task = Task.detached { [transfer, self] in
+          var iterator = transfer
+          do {
+            while await iterate() {
+              if let element = try await iterator.next() {
+                emit(.success(element))
+              } else {
+                emit(.success(nil))
+              }
+            }
+          } catch {
+            emit(.failure(error as! Failure))
+          }
+        }
+        state.withLock { state in
+          precondition(state.iteratingTask.isStarting)
+          state.iteratingTask = .running(task)
+        }
+      }
+      let result: Result<Element?, Failure> = await withTaskCancellationHandler {
```

Comment:
i haven't actually tested to verify, but i suspect the cancel handler body is currently subject this [executor hopping issue](https://github.com/swiftlang/swift/pull/80753) if the isolated `actor` parameter isn't captured in the closure. i'm not sure it would be a correctness issue either way, but presumably the isolation should be forwarded so there's no need to risk hopping for no real reason – potentially just as a defensive measure to reduce the risk of surprising behavior.